rust-lang/cargo
# [experiment] `patch` with patch files

### What does this PR try to resolve?

An unstable `-Zpatch-files` flag is added to support patching your dependencies with patch files, with the patchtool you configure in `[patchtool]` config table.

This is one experiment based on <https://github.com/rust-lang/rfcs/pull/3177> that the team maybe be interested in.
See also #4648.

### How should we test and review this PR?

Apparently commit by commit. You could start from test to understand the current user experience of it.

Some design decisions in this implementation:

* Every patch is settled and applied before the dependency resolution. This implies
  * No re-resolve.
  * Cargo need to fetch a certain version of package beforehand.
    This is currently done by requiring an exact version requirement `=`,
    so Cargo knows what to download.
  * Due to the nature of patches applied before dependency resolution,
    we could patch whatever we want including dependencies and `package.version`.
  * The `Summary` registered to `PacakgeRegistry` is the source after patching.
* `{ patches = [<path>] }` field is only available for dependencies under `[patch]` table,
  and only allows patching registry dependencies for now.
* A new source `PatchedSource` that fetches and applies patches to the original source code during `block_until_ready`.
   * Introduce a new protocol `patched+`, which compose the SourceId to patched.
   * Tracked what was patched in URL query string, for example patching serde@1.0.0 is like:
     ```
     patched+registry+https://github.com/rust-lang/crates.io-index?name=serde&version=1.0.0&patch=/path/to/my-bugfix.patch
     ```
     This SourceId representation also appears in Cargo.lock and Package ID Spec.
* `[patchtool]` table in Config to control.
  I imagine it would be something like git [`mergetool`](https://git-scm.com/docs/git-mergetool) for registering tools,
  though at this moment it only accepts one via `path`.
* Cache of patched source resides in `$CARGO_HOME/patched-src/<hash-of-original-source>/<pkg>-<version>/<cksum-of-patches>`.
  Patched source rebuilds whenever content of patch files or the underlying source change.
  
  
You can find some more documentation under `src/cargo/sources/patched.rs`.

### Additional information

I don't think patching before dependency resolution is ergonomic.
Your patched source code would be stuck in time and you need to eagerly update and rebuild them.
Some patches may be safely applied across major versions.
However with the current design you need to specify the same patch twice.

We may like to revisit patching during the resolution,
and reconsider the work of re-resolving dependencies.
That is a more user-friendly way for patching with patch files.
